### PR TITLE
ImGui Setup

### PR DESCRIPTION
**Pull request by hillj35**
Sets up basic ImGui stuff

- adds an `AddImguiFunction` function to `Renderer`
- vulkan backend handles properly rendering all windows added with `AddImguiFunction`
- add `ImGui` to vcpkg, chose version 1.90.2 because 1.90.3 changes some things with the vulkan implementation. I'd prefer newer ImGui version but it hasn't been added to vcpkg yet, can change it when it gets updated, but don't want it breaking if vcpkg grabs a new version.

Screenshot from 2024-04-13 22-17-04

### CODE REVIEW

File: src/ic_renderer.h [hunk 1]
```diff
@@ -16,13 +19,19 @@ namespace IC {
...
     class Renderer {
     public:
-        Renderer(RendererConfig config) { window = config.window; };
+        Renderer(RendererConfig config) { window = config.window; }
         ~Renderer() = default;
 
-        virtual void DrawFrame() = 0;
+        void AddImguiFunction(ImGuiFunction function) { imGuiFunctions.push_back(function); }
+        void RemoveImguiFunction(ImGuiFunction function) {
+            imGuiFunctions.erase(std::find(imGuiFunctions.begin(), imGuiFunctions.end(), function));
```
Comment (DevWizaad) [inline src/ic_renderer.h]: Does the order of the imGuiFunctions vector matter? If not can we use [std::erase](https://en.cppreference.com/w/cpp/container/vector/erase2) instead? It'll be more efficient here

File: src/vulkan/vulkan_initializers.cpp [hunk 2]
```diff
@@ -163,4 +163,26 @@ namespace IC {
...
+        initInfo.DescriptorPool = descriptorPool;
+        initInfo.MinImageCount = 3;
+        initInfo.ImageCount = 3;
+        initInfo.UseDynamicRendering = true;
+        initInfo.MSAASamples = VK_SAMPLE_COUNT_1_BIT;
+        initInfo.ColorAttachmentFormat = imageFormat;
+
+        ImGui_ImplVulkan_Init(&initInfo, VK_NULL_HANDLE);
+        ImGui_ImplVulkan_CreateFontsTexture();
+        ImGui_ImplVulkan_DestroyFontsTexture();
```
Comment (DevWizaad) [inline src/vulkan/vulkan_initializers.cpp]: What are these two functions doing? Creating the Destroying the Fonts texture?

File: src/vulkan/vulkan_renderer.cpp [hunk 3]
```diff
@@ -38,6 +46,14 @@ namespace IC {
     }
 
     void VulkanRenderer::DrawFrame() {
+        ImGui_ImplVulkan_NewFrame();
+        ImGui_ImplGlfw_NewFrame();
+        ImGui::NewFrame();
+        for (auto function : imGuiFunctions) {
+            function();
+        }
+        ImGui::Render();
```
Comment (DevWizaad) [inline src/vulkan/vulkan_renderer.cpp:55]: From what I remember ImGui wants to be rendered after everything else. Is that still the case or does it not matter if we do it first here?

File: src/ic_renderer.h (hunk 1 above)
Comment (hillj35) [reply src/ic_renderer.h]: I don't think order matters? I switched to erase instead

thanks, I figured there was probably a better way to do this lol

File: src/vulkan/vulkan_initializers.cpp (hunk 2 above)
Comment (hillj35) [reply src/vulkan/vulkan_initializers.cpp]: Apparently neither of these are needed as it's done behind the scenes now 
```
2023-11-10: *BREAKING CHANGE*: Removed parameter from ImGui_ImplVulkan_CreateFontsTexture(): backend now creates its own command-buffer to upload fonts.
//              *BREAKING CHANGE*: Removed ImGui_ImplVulkan_DestroyFontUploadObjects() which is now unecessary as we create and destroy those objects in the backend.
//              ImGui_ImplVulkan_CreateFontsTexture() is automatically called by NewFrame() the first time.
//              You can call ImGui_ImplVulkan_CreateFontsTexture() again to recreate the font atlas texture.
//              Added ImGui_ImplVulkan_DestroyFontsTexture() but you probably never need to call this.
```

**Review by DevWizaad — APPROVED** [review]

File: src/vulkan/vulkan_renderer.cpp [hunk 4]
```diff
@@ -1,8 +1,9 @@
 #include "vulkan_renderer.h"
 
-#include "vulkan_util.h"
```
Comment (Colbydude) [inline src/vulkan/vulkan_renderer.cpp]: Went ahead and fixed this, but I think the pattern for includes we're going for is:

```cpp
#include "my_header.h"

#include <project_include_level_includes.h>

#include "project_same_dir_includes.h"

#include <third_party_lib_includes.h>

#include <std_libraries>
```

File: cmake/icengine-config.cmake.in [hunk 5]
```diff
@@ -4,6 +4,7 @@ if (NOT TARGET IC::ICEngine)
     include(CMakeFindDependencyMacro)
     find_dependency(glfw3 CONFIG REQUIRED)
     find_dependency(glm CONFIG REQUIRED)
+    find_dependency(imgui CONFIG REQUIRED)
```
Comment (Colbydude) [inline cmake/icengine-config.cmake.in:7]: Fixed this, but don't forget about this file too!

File: src/vulkan/vulkan_renderer.h [hunk 6]
```diff
@@ -13,19 +14,25 @@ namespace IC {
...
         SwapChain _swapChain;
         PipelineManager _pipelineManager{};
-        DescriptorAllocator _descriptorAllocator{};
+        DescriptorAllocator _meshDescriptorAllocator{};
+        DescriptorAllocator _imGuiDescriptorAllocator{};
 
         std::vector<MeshRenderData> _renderData;
         std::vector<VkCommandBuffer> _cBuffers;
+
+        // render functions for mac support
```
Comment (Colbydude) [inline src/vulkan/vulkan_renderer.h:36]: :eyes:

**Review by Colbydude — APPROVED** [review]

**Pull request merged**